src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpFormatter.cs

```diff
@@ -99,46 +83,35 @@ internal class CSharpFormatter
             return mappedEdits;
         }
 
-        public int GetCSharpIndentation(FormattingContext context, int projectedDocumentIndex, CancellationToken cancellationToken)
+        public async Task<int> GetCSharpIndentationAsync(FormattingContext context, int projectedDocumentIndex, CancellationToken cancellationToken)
         {
-            if (context is null)
-            {
-                throw new ArgumentNullException(nameof(context));
-            }
-
-            // Add a marker at the position where we need the indentation.
             var changedText = context.CSharpSourceText;
-            var marker = $"{context.NewLineString}#line default{context.NewLineString}#line hidden{context.NewLineString}";
-            changedText = changedText.WithChanges(new TextChange(TextSpan.FromBounds(projectedDocumentIndex, projectedDocumentIndex), marker));
+            var marker = "/*__marker__*/";
+            var markerString = $"{context.NewLineString}{marker}{context.NewLineString}";
+            changedText = changedText.WithChanges(new TextChange(new TextSpan(projectedDocumentIndex, 0), markerString));
             var changedDocument = context.CSharpWorkspaceDocument.WithText(changedText);
 
-            // Get the line number at the position after the marker
-            var line = changedText.Lines.GetLinePosition(projectedDocumentIndex + marker.Length).Line;
+            var formattedDocument = await CodeAnalysis.Formatting.Formatter.FormatAsync(changedDocument, cancellationToken: cancellationToken);
```

Comment:
So for every mapping we format the entire document & retrieve the formatted text? This will super quickly fall over, especially with the ToStringing every time as you mentioned above.

Reply:
Yes :(. This is definitely not a long term fix. But I thought this'll be acceptable for full formatting until the issue on Roslyn is fixed. I will have to find a different way for ontype formatting though.

Reply:
@jinujoseph is [this](https://github.com/dotnet/roslyn/issues/49492) something we could have investigated to understand the cost on the Roslyn side?

Reply:
I also have other ideas on how to avoid per-mapping formatting but those require a slightly larger re-design and I wanted to have some fix out before taking more time to fix that.